src/mpid/common/genq/mpidu_genq_shmem_types.h

```diff
@@ -0,0 +1,99 @@
+#define MPIDU_GENQI_SHMEM_TYPES_H_INCLUDED
+
+#include "mpl.h"
+#include "mpidu_init_shm.h"
+
+/* Public Interface */
+
+typedef void *MPIDU_genq_shmem_pool_t;
+
+typedef struct MPIDU_genq_shmem_queues_t {
+    void *base;                 /* base address for the cells */
+    void *ptr;                  /* pointer to shmem allocation */
+} MPIDU_genq_shmem_queues_t;
+
+static inline void *MPIDU_genq_shmem_pool_base(MPIDU_genq_shmem_pool_t pool);
+
+int MPIDU_genq_shmem_queues_alloc(int num_of_queues, MPIDU_genq_shmem_queues_t * queues);
+int MPIDU_genq_shmem_queues_free(MPIDU_genq_shmem_queues_t queues);
+int MPIDU_genq_shmem_queue_init(MPIDU_genq_shmem_queues_t queues, int queue_idx, int flags);
+
+static inline int MPIDU_genq_shmem_queue_dequeue(MPIDU_genq_shmem_queues_t queues,
+                                                 int queue_idx, void **cell);
+static inline int MPIDU_genq_shmem_queue_enqueue(MPIDU_genq_shmem_queues_t queues,
+                                                 int queue_idx, void *cell);
+
+#define MPIDU_GENQ_SHMEM_QUEUE_TYPE__MPSC MPIDU_GENQ_SHMEM_QUEUE_TYPE__NEM_MPSC
+
+typedef enum {
+    MPIDU_GENQ_SHMEM_QUEUE_TYPE__SERIAL,
+    MPIDU_GENQ_SHMEM_QUEUE_TYPE__INV_MPSC,
+    MPIDU_GENQ_SHMEM_QUEUE_TYPE__NEM_MPSC,
+} MPIDU_genq_shmem_queue_type_e;
+
+/* Internal structure and declarations */
+
+typedef struct genqi_shmem_cell_header {
+    uintptr_t handle;
+    int block_idx;
+    union {
+        struct {
+            uintptr_t next;
+        } serial_queue;
+        struct {
+            uintptr_t prev;     /* used in the tail queue */
+            uintptr_t next;     /* used in the detached header queue */
+        } inverse_queue;
+        struct {
+            MPL_atomic_ptr_t next_m;
+        } nem_queue;
+    } u;
+} genqi_shmem_cell_header_s;
+
+typedef union genqi_shmem_queue {
+    struct {
+        union {
+            uintptr_t s;
+            MPL_atomic_ptr_t m;
+            uint8_t pad[MPIDU_SHM_CACHE_LINE_LEN];
+        } head;
+        union {
+            uintptr_t s;
+            MPL_atomic_ptr_t m;
+            uint8_t pad[MPIDU_SHM_CACHE_LINE_LEN];
+        } tail;
+        unsigned flags;
+    } q;
+    uint8_t pad[3 * MPIDU_SHM_CACHE_LINE_LEN];
+} genqi_shmem_queue_u;
+
+typedef struct genqi_shmem_pool {
+    uintptr_t cell_size;
+    uintptr_t cell_alloc_size;
+    uintptr_t cells_per_proc;
+    uintptr_t num_proc;
+    int rank;
+
+    void *slab;
+    genqi_shmem_cell_header_s *cell_header_base;
+    genqi_shmem_cell_header_s **cell_headers;
+    MPIDU_genq_shmem_queues_t free_queues;
+} genqi_shmem_pool_s;
+
+#define HEADER_TO_CELL(header) \
+    ((char *) (header) + sizeof(genqi_shmem_cell_header_s))
+
+#define CELL_TO_HEADER(cell) \
+    ((genqi_shmem_cell_header_s *) ((char *) (cell) - sizeof(genqi_shmem_cell_header_s)))
+
+/* The handle value is calculated using (address_offset + 1), see MPIDU_genq_shmem_pool.h */
+#define HANDLE_TO_HEADER(base, handle) \
```

Comment:
I don't see what this change is needed.

Reply:
Detach the dependency of queue from pool. I thought you argued for it.